CIP-????/README.md

```diff
@@ -0,0 +1,320 @@
+
+``` typescript
+export type TreeProof =
+  | {
+      // `hash` is the valHash of the Val of the node
+      HashNode: { hash: string; left: TreeProof; right: TreeProof }
+    }
+  | {
+      // `hash` is the rootHash of a node, a branch etc.
+      NodeHash: { hash: string }
+    }
+```
+
+
+This structure regenerates the exact hash of the tree as the whole IntegriTree would be hashed.
+
+In merkle tree the leaf pairs (that contain data) are hashed up to the tree while in IntegriTree the nodes' values (`Val`) are part of the proof.
+For example, for the following Proof
+
+``` typescript
+const proof = HashNode {
+  hash: valHash(Val{ xi: 1, xa: "a", xb: "adam"}),
+  left: HashNode {
+            hash: valHash(Val{ xi: 2, xa: "adam", xb: "b"}),
+            left: emptyHash,
+            right: emptyHash
+        },
+  right: emptyHash
+}
+```
+
+It's easily can be proven that "adam" is in the tree, but "aby" is not, having a `Val { xi: "1", xa: "a", xb: "adam"}` and the Proof
+as the proof's `rootHash(proof)` exactly the same with the rootHash of the whole integritree, and the hash of proof contains the hash of the required proovable `Val`
+
+### Root Hash Calculation of the minimal subtree
+
+The root hash of the minimal subtree use similar function as in the IntegriTree, see the `aiken`'s implementation below:
+
+``` gleam
+pub fn root_hash(root: Proof) -> Hash {
+  when root is {
+    NodeHash { hash } -> hash
+    HashNode { hash, left, right } ->
+      combine_three_hashes(hash, root_hash(left), root_hash(right))
+  }
+}
+```
+
+## Membership and Non-Membership Proofs
+
+IntegriTree provides proofs for both membership and non-membership of elements in the tree:
+
+- **Membership**: If an `x` element is in the tree, there exists at least one node where either `xa = x` or `xb = x`.
+- **Non-Membership**: If an element is not in the tree, there exists a node where `xa < x < xb`.
+
+## Append and Delete Proofs
+
+Proofs are available for appending new elements to the tree and deleting existing elements:
+
+- **Append Proof**: If an element is not in the tree, the proof contains the non-membership (update node's) value `Vu`, the value `Vp` of the node will be the parent of the new appendable node of the new element, and the minimal subtree of the two `Val`'s nodes.
+- **Delete Proof**: If an element is in the tree, the proof contains the two membership (update nodes') values the `Vu1` and `Vu2` (the `(xα, x)` and `(x, xβ)`),  and the value `Vp` the parent node of the last (in level-order) node in the tree, and the minimal subtree create from the nodes of the three values.
+
+
+### Val Types of the Proof
+
+There are two types of Vals are used with the Proof:
+
+- **Updateable Node's (`Nu`'s)**: Nodes' val that are already part of the tree but may have their values updated.
+- **Parent Node's (`Np`'s)**: Nodes' val where a new leaf will be appended to or removed from.
+
+## Completeness Enforcement
```

Comment:
```suggestion
### Completeness Enforcement
```

Reply:
@ilap very important to maintain standard document sections.